Adds lightning message signing/verification/pk_recovery

sr-gi opened the pull request:
Adds message signing, verification and public key recovery.

Signatures are pk-recoverable zbase32-encoded, following:

```
zbase32(SigRec(SHA256(SHA256("Lightning Signed Message:" + msg))))
```

`SigRec` has first byte 31 + recovery id, followed by 64 byte `sig`. 

close #843

TheBlueMatt commented on lightning/Cargo.toml:
```diff
@@ -28,6 +28,7 @@ unstable = []
 
 [dependencies]
 bitcoin = "0.26"
+zbase32 = "0.1.2"
```
Hmm, I'm not a huge fan of depending on a library for something that should be 20-50 LoC, plus it looks like zbase32 doesn't meet our MSRV (hence why CI is failing). Sadly we can't just copy it and fix it since its LGPL, so we'll need to write one.

sr-gi replied:
Got it. I'll add our own implementation of zbase32.

sr-gi replied:
Added a local `zbase32` encoding based on https://github.com/andreasots/base32 in c7f405e

TheBlueMatt commented on lightning/src/util/zbase32.rs:
```diff
@@ -0,0 +1,137 @@
+        ret.push(ALPHABET[(buf[4] & 0x1F) as usize]);
+    }
+
+    if data.len() % 5 != 0 {
```
This if block can be more succinctly written as `ret.truncate((data.len() * 8 + 4) / 5);`

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+fn sigrec_decode(sig_rec: Vec<u8>) -> Result<RecoverableSignature, Error> {
+    let rsig = &sig_rec[1..];
+    let rid = sig_rec[0] as i32 - 31;
+    
```
nit: please `s/    /\t/` this file and drop the trailing spaces on the lines that have them (like this line).

sr-gi replied:
Do you use any linter or code formatter that follows the project formatting convention? Asking to avoid this kind of things in future commits / PRs.

TheBlueMatt replied:
git show will highlight trailing whitespace so it's hard to miss, aside from that, just use tabs. We've looked into rustfmt, but it tends to make a complete unreadable mess, and there aren't many options to tweak it.

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+    #[test]
+    fn test_recover_pk() {
+        let message = "test message";
+        let sig = "d9tibmnic9t5y41hg7hkakdcra94akas9ku3rmmj4ag9mritc8ok4p5qzefs78c9pqfhpuftqqzhydbdwfg7u6w6wdxcqpqn4sj4e73e";
```
Where did you get this test vector? Are there standard test vectors available for this somewhere?

sr-gi replied:
I just made it up. Since there is no spec I don't think there are any standard test vectors, but I can replicate the ones used by c-lightning and lnd so we can make sure those pass at least.

sr-gi replied:
Couldn't find anything standard, added additional tests from https://github.com/ElementsProject/lightning/blob/1275af6fbb02460c8eb2f00990bb0ef9179ce8f3/tests/test_misc.py#L1925-L1938 at 3b52b18

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+use crate::util::zbase32;
```
All public methods and modules need documentation - the `pub fn`s can probably get something simple, but the module here should describe the signing here, note that its a special extension of c-lightning and lnd and not a part of the spec, and maybe link to some documentation on it at c-lightning and/or lnd.

sr-gi replied:
Added docs at 3b52b18

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+use bitcoin::secp256k1::recovery::{RecoverableSignature, RecoveryId};
+use bitcoin::secp256k1::{Error, Message, PublicKey, Secp256k1, SecretKey};
+
+static LN_MESSAGE_PREFIX: &[u8] = "Lightning Signed Message:".as_bytes();
```
`.as_bytes()` in a const isn't stable in our MSRV, but you can use a `b` prefix.
```suggestion
static LN_MESSAGE_PREFIX: &[u8] = b"Lightning Signed Message:";
```

sr-gi replied:
My bad, I actually had it that way but rebased to changed the order of the commits and this must have reset to the old way.

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+    
+}
+
+pub fn verify(msg: &[u8], sig: String, pk: PublicKey) -> bool {
```
`sig` can be an `&str`, not a `String`, no?

sr-gi replied:
Sure

TheBlueMatt replied:
Best to do it in both, I think. In general, taking ownership of something on the heap should only be done if we need it, otherwise it should be a reference.

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+    Ok(zbase32::encode(&sigrec_encode(sig)))
+}
+
+pub fn recover_pk(msg: &[u8], sig: String) ->  Result<PublicKey, Error>{
```
Does this need to be pub?

sr-gi replied:
It may so it can be bound to foreign langs, doesn't it?

TheBlueMatt replied:
Right, my question is what's the use for it - do users have a direct need for this?

sr-gi replied:
Ohh, there is.

lnd's `verifymessage` returns a bool + the recovered public key.

Furthermore, in the Eye of Satoshi case, we actually use this to authenticate users. All requests are signed but public keys are only shared when the user is registered, so EC recovery is used elsewhere.

TheBlueMatt replied:
Ah, ok! May be useful to rename this function, then - something that includes the word "verify" in it, indicating that the message is valid if the public key matches. Needs docs either way.

sr-gi replied:
Checking this further I think there's a need for the three of them to be public.

`lnd`'s implementation verifies the message by checking if the recovered public key matches the `id` of any of the nodes in the node's database.

`c-lightning`'s allows you to provide a public key to verify. If you do, then it verifies the signature agains the provided key, otherwise it does the same as `lnd`'s, and checks it agains the known nodes.

How you check is the recovered key is part of a set of known nodes looks implementation dependant to me, so it feels making this public makes the most sense. Furthermore, in order to allow users to check this agains something different than the nodes in the network (e.g. to allow ephemeral identities or identities created from key tweaking) this would be required. This is the exact use case for `teos`.

TheBlueMatt replied:
I kinda like having the `verify` one, its certainly a much, much safer API, whereas this API its possible to forget to compare the public key against something known-good. Obviously if there's a use for it, then we should keep this, but having the server version feels nice too.

TheBlueMatt reviewed (commented):
There are a number of issues in the zbase32 implementation, but I went ahead and fixed those while writing new tests at https://git.bitcoin.ninja/?p=rust-lightning;a=shortlog;h=refs/heads/fix-zbase32 so you can just pull the commits off that branch.

There's a few comments in the message_signing stuff too, but mostly just documentation and API questions.

sr-gi commented on lightning/src/util/zbase32.rs:
```diff
@@ -0,0 +1,137 @@
+const ALPHABET: &'static [u8] = b"ybndrfg8ejkmcpqxot1uwisza345h769";
+
+pub fn encode(data: &[u8]) -> String {
+    let mut ret = Vec::with_capacity((data.len() + 3) / 4 * 5);
```
I'm not sure either if I'm honest. I checked several base32 / zbase32 implementations and they did it rather differently. My bad for assuming it was right.

Here's c-lightning's, which does not seem to be the same either: https://github.com/ElementsProject/lightning/blob/6f870dfe399461fc556da4a8dea29eeef62b91ad/lightningd/signmessage.c#L34

sr-gi commented on lightning/src/util/zbase32.rs:
```diff
@@ -0,0 +1,137 @@
+
+    let data = data.as_bytes();
+    let output_length = data.len() * 5 / 8;
+    let mut ret = Vec::with_capacity((output_length + 4) / 5 * 5);
```
This one is closer:

https://github.com/ElementsProject/lightning/blob/6f870dfe399461fc556da4a8dea29eeef62b91ad/lightningd/signmessage.c#L59

Care to share how did you came up with the bounds?

TheBlueMatt replied:
Look at the loop - it pushes a fixed number of elements per fixed number of elements, rounded up.

sr-gi commented on lightning/src/util/zbase32.rs:
```diff
@@ -0,0 +1,137 @@
+    21, 9, 10, -1, 11, 2, 16, 13, 14, 4, 22, 17, 19, -1, 20, 15, 0, 23,
+];
+
+pub fn decode(data: &str) -> Result<Vec<u8>, &'static str> {
```
Sure, that'll work too.

ariard commented on lightning/src/util/zbase32.rs:
```diff
@@ -0,0 +1,142 @@
+
+/// Decodes a zbase32 string to the original bytes, failing if the string was not encoded by a
+/// proper zbase32 encoder.
+pub fn decode(data: &str) -> Result<Vec<u8>, ()> {
```
This work well `pub(super)`, no need to make them public ?

sr-gi replied:
I made it public so it can be easily bound with methods that import `lightning` instead of working within the same repo for the bindings, like the Python bindings do. Happy do discuss it though.

TheBlueMatt replied:
Does that mean you need zbase32 for other uses, or just generally you think people may need zbase32 for other things and want it via bindings?

sr-gi replied:
It may not be needed for anything but signing, so I guess we could make it private.

ariard commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,90 @@
+}
+
+
+pub fn sign(msg: &[u8], sk: SecretKey) -> Result<String, Error> {
```
Should they get a document how to use them if they're part of the API ?

sr-gi replied:
Added docs at 3b52b18

ariard reviewed (commented):
Concept ACK.

> I just made it up. Since there is no spec I don't think there are any standard test vectors, but I can replicate the ones used by c-lightning and lnd so we can make sure those pass at least

If you feel it, maybe just propose a spec against the BOLT/BIPs ? It should be easy to land if you have test vectors and other implementations already have support.

sr-gi commented:
Addressed formatting issue, nits, docs missing and added additional tests.

> If you feel it, maybe just propose a spec against the BOLT/BIPs ? It should be easy to land if you have test vectors and other implementations already have support.

Wouldn't mind proposing adding this to the specs, but not sure where it would fit best. Any suggestions?

TheBlueMatt commented:
> Wouldn't mind proposing adding this to the specs, but not sure where it would fit best. Any suggestions?

Presumably a new BOLT? I dunno.

Note that CI is failing because the `zbase32` module doesn't have documentation (ie via `//!` comments at the top of `zbase32.rs`).

TheBlueMatt commented:
I don't think we should have to wait on a new BOLT for this, though one would certainly be nice. I think the only thing still needed is some module-docs on zbase32.

sr-gi commented:
I've changed zbase32 to to pub(crate) base on https://github.com/rust-bitcoin/rust-lightning/pull/844#discussion_r602364102. If you feel there may be an additional use case that would require this to be public I'm happy to rebase and add docs.

TheBlueMatt commented on lightning/src/util/mod.rs:
```diff
@@ -15,9 +15,11 @@ pub(crate) mod fuzz_wrappers;
 
 pub(crate) mod byte_utils;
 pub(crate) mod chacha20;
+pub(crate) mod zbase32;
```
Oops, the fuzztarget mode needs it, see CI.
```suggestion
#[cfg(feature = "fuzztarget")]
pub mod zbase32;
#[cfg(not(feature = "fuzztarget"))]
pub(crate) mod zbase32;
```

sr-gi replied:
Added it in cf384a3 and changed ``zbase32::{encode, decode}`` to pub so they can be used in fuzzing.

TheBlueMatt commented on lightning/src/lib.rs:
```diff
@@ -31,6 +31,8 @@
 
 extern crate bitcoin;
+
+
```
nit: excess spaces?

sr-gi replied:
Oops, didn't see that one.

TheBlueMatt commented on lightning/src/util/mod.rs:
```diff
@@ -15,10 +15,14 @@ pub(crate) mod fuzz_wrappers;
+#[cfg(feature = "fuzztarget")]
+pub mod zbase32;
 #[cfg(not(feature = "fuzztarget"))]
+pub(crate) mod zbase32;
```
The compile error in CI is that `poly1305` is no longer `#[cfg(not(feature = "fuzztarget"))]` (ie not compiled in fuzzing). You need to duplicate the cfg line, above.

valentinewallace commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,145 @@
+
+/// Verifies a message was signed by a PrivateKey that derives to a given PublicKey, given a message, a signature,
+/// and the PublicKey.
+pub fn verify(msg: &[u8], sig: &str, pk: PublicKey) -> bool {
```
I tested this with the sample and `lnd`'s telling me the message is invalid (the pubkey's correct): https://i.imgur.com/lizoFib.png

Haven't tested further. 
Instructions to reproduce/test further (I used Polar but feel free to use whatever you want):
* clone this branch of the sample: https://github.com/valentinewallace/ldk-sample/tree/for-testing-signing (it uses a branch of `rust-lightning` that's just your branch + one commit to expose a struct that's needed)
* I used Polar: https://github.com/jamaljsr/polar by cloning and running `yarn dev`
* start a new test network in Polar and add an `lnd` node
* start the sample: `cargo run polaruser:polarpass@127.0.0.1:18443 ./ 9738 regtest`
* in the sample CLI: `connectpeer <lnd node>`, where `lnd node`'s found under `alice > Connect > P2P External`
* in the sample CLI: `signmessage` (the message is hardcoded, see: https://github.com/valentinewallace/ldk-sample/blob/for-testing-signing/src/cli.rs#L378)
* copy this output, then open a terminal for the lnd node under `alice > Actions > Terminal > Launch`
* `lncli verifymessage --msg abcdefghijklmnopqrstuvwxyz1234567890 --sig <sig you just copied>`

Also added `verifymessage` to the sample: https://github.com/valentinewallace/ldk-sample/blob/for-testing-signing/src/cli.rs#L390 if you wanna test that

sr-gi replied:
> I tested this with the sample and lnd's telling me the message is invalid (the pubkey's correct): https://i.imgur.com/lizoFib.png

Wouldn't this mean the two nodes are not properly connected (or the node that originated the signature is not on the verifiers database?). I'm wondering since you point that the recovered public key is correct, so the signature is actually valid.

From [lnd docs](https://api.lightning.community/#verifymessage):

> VerifyMessage verifies a signature over a msg. The signature must be zbase32 encoded and signed by an active node in the resident node's channel database. In addition to returning the validity of the signature, VerifyMessage also returns the recovered pubkey from the signature.

sr-gi replied:
@valentinewallace I double checked this against `lnd` and `c-lightning`. Both return:

```
{
   "pubkey": our_public_key,
   "verified": false
}
```

However, in `c-lightning`, if our public key is passed as optional parameter then the message is verified.

Checking [c-lightning's docs](https://lightning.readthedocs.io/lightning-checkmessage.7.html), we can find the following:

> As a special case, if pubkey is not specified, we will try every known node key (as per listnodes), and verification succeeds if it matches for any one of them.

And, If I check `listnodes`, the list is empty (even though the node is one of our peers), so I'm guessing that's the root of the issue. The pubkey is correctly recovered in both cases, but the node seems not to be part of neither of the peers known nodes, so it fails validation agains known nodes.

sr-gi replied:
Furthermore, querying `describegraph` in lnd's node does not show our testing node, which I'd say backs the aforementioned claim.

I'm not sure why that's happening though. Any thoughts?

sr-gi replied:
Asked around #c-lightning on IRC, looks like a channel announced is required to show-up in `listnodes`, and I'm assuming the same may apply to `describegraph` in `lnd`:

```
20:12 <sr_gi> Hey guys, what is needed for a node to show in listnodes? I was doing some testing using Polar and
connecting a c-lightning node to a rust-lightning node and even though they were connected the peer did no show up
in the list. Is a channel required or something on those lines?
20:14 <@cdecker> An announced channel is required, yes
```

valentinewallace replied:
As discussed offline, this seems like a reasonable explanation for why lnd's giving `verified: false` :)

valentinewallace reviewed (commented):
Sorry to nit you to death on grammar, wouldn't be so pressed if it weren't public docs.

Lmk if the testing instructions are unclear or if I'm in the wrong there

sr-gi commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,145 @@
+//!
+//! Note this is not parts of the specs, but follows lnd's signing and verifying protocol, which can is defined as follows:
+//!
+//! signature = zbase32(SigRec(sha256d(("Lightning Signed Message:" + msg)))
```
No, there's not. Would it be clearer as `"Lightning Signed Message:"+msg`? I formatted the Pythonic's way but didn't realise that may be an issue.

sr-gi commented:
> Sorry to nit you to death on grammar, wouldn't be so pressed if it weren't public docs.

No worries, I actually appreciated it.

Addressed the grammar issues you pointed out and a couple more related I found.

TheBlueMatt commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,140 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Ligthning message signing and verification lives here. These tools can be used to sign messages using the node's
```
Lightning, transpose the t and h

sr-gi replied:
😅

ariard commented on lightning/src/util/message_signing.rs:
```diff
@@ -0,0 +1,140 @@
+use bitcoin::secp256k1::recovery::{RecoverableSignature, RecoveryId};
+use bitcoin::secp256k1::{Error, Message, PublicKey, Secp256k1, SecretKey};
+
+static LN_MESSAGE_PREFIX: &[u8] = b"Lightning Signed Message:";
```
Bind message to a lightning signing context to avoid replay of signed data in another context with same key material ? Though w.r.t to this concern our message API signing is safe as this is always committed by the implementation and not an argument caller, not sure if it's worthy to document.

Do we have risks of harmful API misuses we should document ? I think we're good, can't foresee real ones.

sr-gi replied:
Yeah, that's the purpose of the prefix AFAIK. I can read a comment there stating the purpose of the prefix if you feel it's necessary.

> Do we have risks of harmful API misuses we should document ? I think we're good, can't foresee real ones.

I cannot see any either.

ariard reviewed (commented):
Code Review ACK 7bcf5a1
